### PR TITLE
Update handlebars AST

### PR DESCRIPTION
**Pull request by marcioj**
As requested by @mmun in #290

### CODE REVIEW

**Comment by mmun** [comment]
You should rename the `isHelper` param from sexpr to mustache. Sexprs are always helpers now.

File: packages/htmlbars-compiler/lib/hydration-opcode-compiler.js [hunk 1]
```diff
@@ -121,19 +114,17 @@ HydrationOpcodeCompiler.prototype.closeElement = function() {
...
 
   var morphNum = this.morphNum++;
   var start = this.currentDOMChildIndex;
   var end = this.currentDOMChildIndex;
   this.morphs.push([morphNum, this.paths.slice(), start, end, mustache.escaped]);
 
-  if (isHelper(sexpr)) {
-    prepareSexpr(this, sexpr);
+  if (isHelper(mustache)) {
+    prepareCommonOpcodes(this, mustache);
```
Comment (mmun) [inline packages/htmlbars-compiler/lib/hydration-opcode-compiler.js]: I'd prefer to just expand this out to the 3 calls so that it stays at the same abstraction as the AST.

File: packages/htmlbars-compiler/lib/hydration-opcode-compiler.js [hunk 2]
```diff
@@ -233,8 +222,13 @@ HydrationOpcodeCompiler.prototype.pushMorphPlaceholderNode = function() {
   this.comment();
 };
 
+HydrationOpcodeCompiler.prototype.MustacheStatement = function(mustache) {
```
Comment (mmun) [inline packages/htmlbars-compiler/lib/hydration-opcode-compiler.js]: Why is this here? Seems strange to have both this AND `HydrationOpcodeCompiler.prototype.mustache`, no?

File: packages/htmlbars-syntax/tests/parser-node-test.js [hunk 3]
```diff
@@ -286,42 +286,42 @@ test("Node helpers", function() {
 test("Tokenizer: MustacheStatement encountered in tagName state", function() {
   var t = "<input{{bar}}>";
   astEqual(t, b.program([
-    b.element('input', [], [ b.sexpr(b.path('bar')) ])
+    b.element('input', [], [ b.mustache(b.path('bar')) ])
```
Comment (mmun) [inline packages/htmlbars-syntax/tests/parser-node-test.js]: Hm, these nodes really want to be their own AST node. Something like `ElementModifier`.

File: packages/htmlbars-compiler/lib/hydration-opcode-compiler.js (hunk 2 above)
Comment (marcioj) [inline packages/htmlbars-compiler/lib/hydration-opcode-compiler.js]: This method runs in the context of `ConcatStatement` like `<div foo="something {{bar}}">`. Due to AST update, now it is generated `concat(text("something"), mustache("bar"))` instead of  `concat(text("something"), sexp("bar"))`, so `MustacheStatement` is visited. 
I don't have a deep understanding of what is happening, but I believe that mustaches in attributes behaves different from these outside (Morphs vs AttrMorphs), so using the mustache method in that case could generate a different behavior.

**Comment by mmun** [comment]
:+1: I'm going to defer merging this until the reactive rerending stuff is merged.

File: packages/htmlbars-syntax/tests/parser-node-test.js (hunk 3 above)
Comment (marcioj) [inline packages/htmlbars-syntax/tests/parser-node-test.js]: Sorry, not sure if I get the idea. Do you want to group the `tag`, `attributes` and `helpers` property of `ElementNode` in a node called `ElementModifier`?

Comment (mmun) [inline packages/htmlbars-syntax/tests/parser-node-test.js]: In another PR I want to rename `ElementNode.helpers` &rarr; `ElementNode.modifiers` and change them from `MustacheStatement` to `ElementModifier`.

**Comment by mmun** [comment]
I'm going to tweak your PR and merge it soon, now that Handlebars 3.0.0 has been released.

**Comment by mmun** [comment]
Merged manually in 4f089543c470c4a73f9b068a2dca867d8ce15c53.

**Comment by marcioj** [comment]
Glad to help!
